src/platform/pcp.c

```diff
@@ -0,0 +1,574 @@
+        } PEER;
+    };
+
+} PCP_RESPONSE;
+
+#pragma pack(pop)
+
+#define SIZEOF_THROUGH_FIELD(type, field) \
+    (FIELD_OFFSET(type, field) + sizeof(((type *)0)->field))
+
+const uint16_t PCP_MAP_REQUEST_SIZE = SIZEOF_THROUGH_FIELD(PCP_REQUEST, MAP.SuggestedExternalIpAddress);
+const uint16_t PCP_PEER_REQUEST_SIZE = SIZEOF_THROUGH_FIELD(PCP_REQUEST, PEER.RemotePeerIpAddress);
+
+const uint16_t PCP_MAP_RESPONSE_SIZE = SIZEOF_THROUGH_FIELD(PCP_RESPONSE, MAP.AssignedExternalIpAddress);
+const uint16_t PCP_PEER_RESPONSE_SIZE = SIZEOF_THROUGH_FIELD(PCP_RESPONSE, PEER.RemotePeerIpAddress);
+
+//
+// Main structure for PCP
+//
+typedef struct CXPLAT_PCP {
+
+    void* ClientContext;
+    CXPLAT_PCP_CALLBACK_HANDLER ClientCallback;
+
+    uint32_t GatewayCount;
+
+    _Field_size_(GatewayCount)
+    CXPLAT_SOCKET* GatewaySockets[0];
+
+} CXPLAT_PCP;
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatPcpInitialize(
+    _In_ CXPLAT_DATAPATH* Datapath,
+    _In_ void* Context,
+    _In_ CXPLAT_PCP_CALLBACK_HANDLER Handler,
+    _Out_ CXPLAT_PCP** NewPcpContext
+    )
+{
+    CXPLAT_PCP* PcpContext = NULL;
+    uint32_t PcpContextSize;
+    QUIC_ADDR* GatewayAddresses = NULL;
+    uint32_t GatewayAddressesCount;
+
+    QUIC_STATUS Status =
+        CxPlatDataPathGetGatewayAddresses(
+            Datapath,
+            &GatewayAddresses,
+            &GatewayAddressesCount);
+    if (QUIC_FAILED(Status)) {
+        goto Exit;
+    }
+    CXPLAT_DBG_ASSERT(GatewayAddresses != NULL);
+    CXPLAT_DBG_ASSERT(GatewayAddressesCount != 0);
+
+    PcpContextSize = sizeof(CXPLAT_PCP) + (GatewayAddressesCount * sizeof(CXPLAT_SOCKET*));
+    PcpContext = (CXPLAT_PCP*)CXPLAT_ALLOC_PAGED(PcpContextSize, QUIC_POOL_PCP);
+    if (PcpContext == NULL) {
+        QuicTraceEvent(
+            AllocFailure,
+            "Allocation of '%s' failed. (%llu bytes)",
+            "CXPLAT_PCP",
+            PcpContextSize);
+        Status = QUIC_STATUS_OUT_OF_MEMORY;
+        goto Exit;
+    }
+
+    CxPlatZeroMemory(PcpContext, PcpContextSize);
+    PcpContext->ClientContext = Context;
+    PcpContext->ClientCallback = Handler;
+    PcpContext->GatewayCount = GatewayAddressesCount;
+
+    for (uint32_t i = 0; i < GatewayAddressesCount; ++i) {
+        QuicAddrSetPort(&GatewayAddresses[i], CXPLAT_PCP_PORT);
+        Status =
+            CxPlatSocketCreateUdp(
+                Datapath,
+                NULL,
+                &GatewayAddresses[i],
+                PcpContext,
+                CXPLAT_SOCKET_FLAG_PCP,
+                &PcpContext->GatewaySockets[i]);
+        if (QUIC_FAILED(Status)) {
+            goto Exit;
+        }
+    }
+
+    *NewPcpContext = PcpContext;
+    Status = QUIC_STATUS_SUCCESS;
+
+Exit:
+
+    if (QUIC_FAILED(Status)) {
+        if (PcpContext != NULL) {
+            CxPlatPcpUninitialize(PcpContext);
+        }
+    }
+
+    if (GatewayAddresses != NULL) {
+        CXPLAT_FREE(GatewayAddresses, QUIC_POOL_DATAPATH_ADDRESSES);
```

Comment:
I wonder if we should have a function in the datapath to free this. There are not many other places in code where we depend on the caller to know how to free the memory. It just gets messy.

Reply:
What if I put the function declaration in an internal header (not quic_datapath.h)?

Reply:
In all honesty it doesn't really matter, just something I noticed. Theres probably other places we do this, theyre just rarer. This is fine.